メッセージ送信の非同期通信機能

Hiro3nori3 opened the pull request.

Commit:
message.jsの作成

Commit:
messages_controllerにJSONのフォーマットを追加

Commit:
create.json.jbuilderの作成

Commit:
messages_controllerの修正

・if文が消えていた

Commit:
viewの修正

・htmlの指定がないとエラーがでてしまうため

Commit:
formData

Hiro3nori3 commented:
[![Screenshot from Gyazo](https://gyazo.com/1aaa708f50c9963a544cbcad5835e48b/raw)](https://gyazo.com/1aaa708f50c9963a544cbcad5835e48b)

Hiro3nori3 commented:
[![Screenshot from Gyazo](https://gyazo.com/ce5afb6a92ad5df95835b9acef9dec92/raw)](https://gyazo.com/ce5afb6a92ad5df95835b9acef9dec92)

Commit:
create.json.jbuilderの修正

Commit:
messageの部分テンプレート化

Commit:
ja.ymlの編集

フォームを送信した際に、英語の”Send”から日本語で「登録する」という表記に変わってしまうのを防ぐため

Commit:
message.jsの編集

メッセージをHTMLで描画したり、自動スクロールの機能を追加した

Hiro3nori3 commented:
現在、なぜかmessagesテーブルのデータがない状態で、新しくメッセージをフォームで投稿するとjsのイベントが発火しないという問題に見舞われています。（２回目以降は通常通り、非同期通信が行われる）

（追記)
原因が判明。message.jsで１回目に存在しない要素に対してJSを使っていたことにより、１回目のデータの送信の際にイベントが発火しなかった模様

Commit:
sidebarの部分テンプレート作成

Hiro3nori3 commented:
[![Screenshot from Gyazo](https://gyazo.com/c5ee26d78877770511a860203b30ff6f/raw)](https://gyazo.com/c5ee26d78877770511a860203b30ff6f)

fumihumi commented:
pullRequest作成の際はdescriptionに、このプルリクエストではどんな機能を実装しているのか記入をして下さい！

fumihumi commented on app/assets/javascripts/message.js:
```diff
@@ -0,0 +1,86 @@
+ $(function() {
+  function buildHTML(message){
+      var image = "";
+      image = `<img src="${message.image.url}" class="large-img">`;
```
message.image.url == nullやundefindだった場合エラーになりませんか？

fumihumi commented on app/assets/javascripts/message.js:
```diff
@@ -0,0 +1,86 @@
+  function buildHTML(message){
+      var image = "";
+      image = `<img src="${message.image.url}" class="large-img">`;
+    if(message.body && message.image.url) {
```
この if 文の条件分岐ですが、全体を見た際に似たような記述が多いです！
今後修正箇所が少なくても済むようにして下さい

fumihumi commented on app/assets/javascripts/message.js:
```diff
@@ -0,0 +1,86 @@
+
+  $('#new_form').on('submit', function(e){
+    e.preventDefault();
+    console.log(this);
```
不要な記述は削除して下さい

fumihumi commented on app/assets/javascripts/message.js:
```diff
@@ -0,0 +1,86 @@
+      processData: false,
+      contentType: false
+    })
+
```
`.done`はajax()へのメソッドチェーンになっているのでこの部分の空行は不要です

fumihumi commented on app/assets/javascripts/message.js:
```diff
@@ -0,0 +1,86 @@
+      var html = buildHTML(data);
+      $('.rightcontent--content').append(html);
+      $('.rightcontent--footer__body-form').val('');
+      $('.image_form_button').val('');
```
```
$('.rightcontent--footer__body-form').val('');
$('.image_form_button').val('');
``` 
この部分はDOMへの探索を行わないで初期化できるはずです！
formDataなどで取得したオブジェクトの再利用を考えましょう

fumihumi commented on app/assets/javascripts/message.js:
```diff
@@ -0,0 +1,86 @@
+
+    .fail(function(){
+      alert('メッセージを入力してください');
+      $('.rightcontent--footer__button-send').prop("disabled", false);
```
doneとfailにそれぞれ記述するのではなく、.alwaysなど利用しましょう

fumihumi commented on app/views/messages/_sidebar.html.haml:
```diff
@@ -0,0 +1,19 @@
+					= link_to group.name ,group_messages_path(group), style: "color:white;"
+				.leftcontent--bottom-group-talk
+					= group.messages.last.try(:body) || "まだメッセージはありません"
+
```
お！tryいいですね！ 👍

fumihumi commented on app/views/messages/create.json.jbuilder:
```diff
@@ -0,0 +1,4 @@
+json.user_name  @message.user.name
+json.day        @message.created_at.strftime("%Y/%m/%d %H:%M")
```
> created_at.strftime("%Y/%m/%d %H:%M")
 この記述はモデルやhelperに切り出しメソッドにしましょう！

view などで呼び出す際はそのメソッドを利用しましょう！

fumihumi commented on config/locales/ja.yml:
```diff
@@ -136,7 +136,7 @@ ja:
       prompt: 選択してください。
     submit:
-      create: 登録する
+      create: Send
```
ja_ymlなのに英語にして大丈夫でしょうか？

Hiro3nori3 replied:
これを直すために英語にしてみたのですが、他にどうすればこれ直すことができますかね...?
[![Screenshot from Gyazo](https://gyazo.com/578187694dea67f48d90b8b53762603f/raw)](https://gyazo.com/578187694dea67f48d90b8b53762603f)

fumihumi replied:
ja.ymlとなっているのに、このファイルに英語を入れるのは、形式としておかしいかなと感じます。
もしこれが悪さをしているのであればその行をコメントにするなどで対応する。

もしくは、View側にてlabelなどを利用して実装できないでしょうか？

Hiro3nori3 replied:
わざわざvalue属性に"Send"と記述していたため、form.reset()が動いてgifのようになってました。そこで、value属性を取り除いて記述したら、うまくいきました！

fumihumi reviewed (commented):
主にrubyファイル、erbのインデントが深く感じます。！できれば半角スペース2個分にしましょう！

Commit:
修正ver.1

・message.jsファイルのインデントをタブに変換した
・不要なconsole.log(this)を削除した
・ajaxのチェーンメソッドである、.done, .failメソッドと$.ajaxの間に空行があったので削除した
・.doneメソッドと.failメソッドで共通した記述を.alwaysに記述した
・formの内容のリセットの記述を変えた
・create.json.jbuilderの@messageの作成日時をメソッドで呼び出すようにした
・ja.ymlを以前の状態に戻した
・message.jsの記述をリファクタリングした

fumihumi commented on app/assets/javascripts/message.js:
```diff
@@ -0,0 +1,45 @@
+			if(message.image.url) {
+				html = html.replace(/<\/ul>/g, `<div class="image"><img src="${message.image.url} class="large-img"><\/div><\/ul>`)
+			}
+		$('.rightcontent--content').append(html);
```
buildHTMLという関数名からだと、この関数内では、あくまでも `HTMLの生成` をしていて、別の箇所で`append`しそうなニュアンスだと思いますがいかがでしょうか？
関数名を変更する、もしくはこの関数では生成したDOMをreturnするのが良いかなと感じます

Hiro3nori3 replied:
そこまで考えていませんでした。関数名をappendMessageにしてみたのですが、どうでしょう？

fumihumi replied:
そうですね、append~~(){}の方がより関数名に沿った処理に感じます！ 👍

fumihumi commented on app/views/messages/_message.html.haml:
```diff
@@ -0,0 +1,9 @@
+%div.rightcontent--content-speaker
+	= message.user.name
+%div.rightcontent--content-timestamps
+	= message.created_at
```
ここって、`message_time `利用しませんか？

Hiro3nori3 replied:
見落としていました。。。

fumihumi commented on app/assets/javascripts/message.js:
```diff
@@ -0,0 +1,45 @@
+				$('.rightcontent--content').animate({scrollTop: $('.rightcontent--content')[0].scrollHeight}, 'fast');
+			 })
+			.fail(function(){
+				alert('メッセージを入力してください2');
```
最後の２って、必要でしょうか？？。

Hiro3nori3 replied:
デバッグのためにつけてました。すみません！

fumihumi commented on app/assets/stylesheets/_sidebar.scss:
```diff
@@ -0,0 +1,64 @@
+
+  &--header{
+    height: 100px;
+    background-color: #253141;
```
複雑なカラーコードは変数にしましょう。

fumihumi commented:
全体的にインデントの数が狂っているように感じます。
→.rb, .hamlに関しては二つ
→javascriptは4 or 2 に統一しましょう。
jsに関してはどちらも利用されているので、好みに応じて設定してください。

Hiro3nori3 commented:
インデントの数に関してなのですが、以前指摘されて、途中からインデントにスペース２つ分のタブを使うようになったため、それ以前のファイルが一部きっちりとインデントがされていません。すべてのファイルのインデントをタブに変えてくれるコマンドなどあればいいのですが、そういったものってあったりしますか？
(１つのファイル内だけであったら、Sublime Textにそういったオプションがありますが...)

Commit:
修正ver.2

・message.jsの関数名を内容がわかりやすいように変えた
・sassファイルの一部の複雑な色を変数で扱うようにした
・_message.html.hamlの時間表記を修正ver.1で加えたヘルパーメソッドで呼び出すようにした
・Sendボタンからvalueのオプションを取り除いた

fumihumi commented on app/assets/stylesheets/_variables.scss:
```diff
@@ -0,0 +1,3 @@
+$leftcontent-body: #2f3e51;
+$leftcontent-header: #253141;
+$rightcontent-footer-button: #38AEF0;
```
👍

fumihumi commented on app/views/messages/_sidebar.html.haml:
```diff
@@ -0,0 +1,19 @@
+				%ul.group--name--box
+					= link_to group.name ,group_messages_path(group), style: "color:white;"
+				.leftcontent--bottom-group-talk
+					= group.messages.last.try(:body) || "まだメッセージはありません"
```
最後の投稿がメッセージの時と画像の時とで、表示が違ったような気がしますが、その点の実装ってされてますか？？

Hiro3nori3 replied:
まだしていません。考えてみたのですが、group.rb内にメソッドとして定義することを考えたのですが、非同期で左のサイドバーに変更するにはやはりjsで実装でしょうか？

fumihumi replied:
そうですね、メッセージの非同期昨日を作成した際に該当Groupのメッセージを変更できると良いかもしれません。
ただ、chatSpaceではその実装までしている方は少ない気がします...w

----------------------
あくまで、今回の場合はViewをrenderする際に下記のような条件分岐をViewのなかでやっていただきたいなぁとおもってコメントしています！

message.lastが画像を持っているのか
→持っていたら’画像が送信されました’
画像ではないけどメッセージがある？
→メッセージの表示。
何もなければ上記のように、
→まだメッセージがありません
という風に変更をするようにします。

Hiro3nori3 replied:
あ、それで大丈夫だったんですね！

本番のchatspaceの動作もみてみたのですが、非同期で表示するようにはなっていませんでした（笑）

fumihumi replied:
> 本番のchatspaceの動作もみてみたのですが、非同期で表示するようにはなっていませんでした（笑）

あら、、チャレンジしてもいいと思いますよ！ 👍

Hiro3nori3 replied:
難しそうなので、また機会があればチャレンジしようと思います苦笑

Hiro3nori3 replied:
[![Screenshot from Gyazo](https://gyazo.com/438b3525284085afe9e3f739ec85ed2d/raw)](https://gyazo.com/438b3525284085afe9e3f739ec85ed2d)

Hiro3nori3 replied:
lastメソッドを使って取ってきたレコードのカラムをとろうとしているのですが、とれません。。。助けていただきたいです。

Hiro3nori3 replied:
Carrierwaveのヘルパーメソッドとか関係ありますか？

fumihumi replied:
[![Screenshot from Gyazo](https://gyazo.com/19579264e8becd161e450df967428ddd/raw)](https://gyazo.com/19579264e8becd161e450df967428ddd)

fumihumi replied:
ぱっとみlastのメッセージのimageを取得できていると思いますよ！

fumihumi replied:
carrierWaveを使ってると画像は↑みたいな感じになりますね！

Hiro3nori3 replied:
ご丁寧にありがとうございます！
なぜか、undefined method `image' for nil:NilClassというエラーが吐かれます。

fumihumi replied:
おお、なるほど。。 
もしかして、今いる`groupの最後のメッセージ`に画像が投稿されていないんじゃないでしょうか？？
→それらを含めた条件分岐をして欲しいです！

dataBaseにimageがあるのに、エラー担っているのであれば別問題ですが。。。
@Hiro3nori3

Hiro3nori3 replied:
ちょっともう一度その辺調べてみます！今日はひとまず、ありがとうございました！
明日にはこれを終わらして、インクリメンタルサーチのところ入れるように頑張ります！

Hiro3nori3 replied:
原因がわかりました。現在グループが２つあるのですが、一方のグループにメッセージがないため、eachの中でgroup.messages.imageをしようとするとundefined method for nilClassになるということでした。

Commit:
修正ver.4

微調整

Commit:
修正ver.5

・左側のサイドバーに①メッセージがあるときとないとき②メッセージがあるとき、それが画像なのかテキストなのかによって表示を変えるような条件分岐を記
述

fumihumi commented on app/views/messages/_sidebar.html.haml:
```diff
@@ -0,0 +1,26 @@
+			%ul.group--name--box
+				= link_to group.name ,group_messages_path(group), style: "color:white;"
+			.leftcontent--bottom-group-talk
+				- if group.messages.present?
```
この`if `はネストしないで書きたいですね！

Hiro3nori3 replied:
OKです！

fumihumi commented on app/views/messages/_sidebar.html.haml:
```diff
@@ -0,0 +1,26 @@
+				= link_to group.name ,group_messages_path(group), style: "color:white;"
+			.leftcontent--bottom-group-talk
+				- if group.messages.present?
+					- group.messages.each do |message|
```
ここってeachする必要ってありますか？？？？
groupに投稿された最後のmessageを表示するだけなので、
 - if group.message.last.~
で取得できそうです

Hiro3nori3 replied:
色々ややこしくしてすみません！

fumihumi commented on app/views/messages/_sidebar.html.haml:
```diff
@@ -0,0 +1,26 @@
+					- group.messages.each do |message|
+						- if message.present? && message.image.present?
+							=	"画像を投稿しました"
+						- if message.present? && !(message.image.present?)
```
ifを続けて書くのではなく、elsifを有効活用しましょう！

Commit:
修正 ver.6

・修正ver.5の修正

Commit:
修正 ver.6-2

・条件分岐が２つしかないのでelsifではなく、elseに変更

fumihumi commented on app/views/messages/_sidebar.html.haml:
```diff
@@ -0,0 +1,23 @@
+				= link_to group.name ,group_messages_path(group), style: "color:white;"
+			.leftcontent--bottom-group-talk
+			- if group.messages.present?
+				- if (group.messages.last.image?)
```
17,18行目において、この部分でif構文を２つ使わずに `and条件`なども利用して if構文を一つにまとめてみましょう！
そっちの方がスッキリすると思います！
17 ~ 23行目あたりをもう少し直してみてください！

Hiro3nori3 replied:
リファクタリング大事ですね。ご指摘ありがとうございます！

fumihumi commented on app/views/messages/index.html.haml:
```diff
@@ -29,23 +13,15 @@
 
 	.rightcontent--footer
-		= form_for [@group,@message], id: {'new_form'} do |f|
+		= form_for [@group,@message], html: {id: 'new_form'} do |f|
```
このidって指定しなくとも、form_forで勝手に作成されませんか？？？？

Hiro3nori3 replied:
form_for便利ですね。。。

Commit:
修正 ver.6-3

修正ver.6-2の修正
・if文のリファクタリング
・form_forのidの記述を削除
・jsも変更

fumihumi commented:
![](http://www.lgtm.in/p/wqjEdbqPp)
